standards/application/messaging-api.md

```diff
@@ -0,0 +1,965 @@
+- [STORE](../standards/core/store.md)
+- [LIGHTPUSH](../standards/core/lightpush.md)
+- [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md)
+
+This API extends the core protocols with additional functionality
+required for peer-to-peer messaging applications, including:
+- Defines a set of events which MAY be used for subscription to track message propagation, node state, and error handling.
+- Defines a set of method calls which MAY be used for invocation of protocol operations.
+- Abstracts new primitives to be accessed in a RESTful manner via [Waku REST API](https://waku-org.github.io/waku-rest-api/).
+
+This document defines the specifications and guidelines necessary for implementing the API,
+ensuring interoperability and consistency across the Waku protocol family.
+
+## Design Requirements
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, 
+“RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+### Definitions
+
+This section defines key terms and concepts used throughout this document.
+Each term is detailed in dedicated subsections to ensure clarity and consistency in interpretation.
+
+`Multiaddr` - a self-describing format for encoding network address of a remote peer as per [libp2p addressing](https://github.com/libp2p/specs/blob/6d38f88f7b2d16b0e4489298bcd0737a6d704f7e/addressing/README.md) specification.
+
+### Motivation
+
+Real-world application development has exposed challenges in achieving comprehensive usability guarantees for core Waku protocols.
+Although recommendations such as [P2P-RELIABILITY](./p2p-reliability.md) have been introduced to enhance protocol robustness,
+the fragmented landscape of guidelines and the requirement to re-implement reliability strategies across diverse use cases remain evident.
+
+A significant challenge has been the lack of a unified abstraction for message exchange.
+Developers require a straightforward mechanism to send,
+receive, and track messages—whether originating from a node or propagated throughout its connected network.
+
+## API design
+
+### Requirements
+
+This API is designed for generic use and ease of implementation across `nwaku`, `js-waku`,
+and traditional `REST` architectures.
+From this point forward,
+relevant code blocks will be provided in `JSON` format, `HTTP` format or by using `TypeScript` syntax.
+
+### Initial configuration
+
+The `Messaging API` is an abstraction layer built upon the basic functionality provided by a node.  
+The configuration definitions that follow omit implementation-specific details,
+focusing exclusively on settings explicitly required by the `Messaging API`.
+
+```typescript
+{
+  mode: "edge" | "service";
+  clusterId: number;
+  shards: number[];
+  storeNodes?: string[];
+  serviceNodes?: string[];
```

Comment:
Would the assumption be that same nodes are used for publishing and receiving (via lightpush and filter)? Since we have storeNodes separate, would it make sense to also have something like `publishNodes`, `listenNodes`?

Reply:
That is my understanding of this field - preferred service nodes, not "the only service nodes to use"

Reply:
renaming to `preferredServiceNodes` as this is original intention and specified in description already

Reply:
https://github.com/waku-org/specs/pull/19/commits/e60bd6f6248c0f7d2c5b967ef7910193b3b394c2